### PR TITLE
Enable mlserver to encode pandas (series and dataframe) from mlflow model output

### PR DESCRIPTION
**Pull request by sakoush**
As per the mlflow inference [signature](https://www.mlflow.org/docs/latest/python_api/mlflow.pyfunc.html#module-mlflow.pyfunc):
```
predict(
  model_input: [pandas.DataFrame, Dict[str, numpy.ndarray], numpy.ndarray]
) -> [numpy.ndarray | pandas.(Series | DataFrame)]
```
predict can return pandas (e.g. in the case of pytorch mnist example).

This PR adds the support to parse pandas if mlflow returns it as inference output.

note: we need to make sure that edge cases are considered.

### CODE REVIEW

File: runtimes/mlflow/tests/test_encoding.py [hunk 1]
```diff
@@ -1,32 +1,38 @@
...
+from mlserver.codecs.numpy import _to_datatype
+
 
 @pytest.mark.parametrize(
     "mlflow_payload",
     [
         np.array([1, 2, 3]),
         {"foo": np.array([1, 2, 3])},
         {"foo": np.array([1, 2, 3]), "bar": np.array([4, 5, 6], dtype=np.float32)},
+        pd.DataFrame([[1, 2, 3]]),
```
Comment (sakoush) [inline runtimes/mlflow/tests/test_encoding.py:19]: @adriangonz what edge cases we need to consider here for pandas?

File: runtimes/mlflow/mlserver_mlflow/encoding.py [hunk 2]
```diff
@@ -1,23 +1,48 @@
...
+    if (
+        isinstance(mlflow_payload, np.ndarray)
+        or isinstance(mlflow_payload, pd.DataFrame)
+        or isinstance(mlflow_payload, pd.Series)
+    ):
+
+        # convert pandas Series or DataFrame to numpy ndarray
+        if isinstance(mlflow_payload, pd.DataFrame):
+            assert isinstance(mlflow_payload, pd.DataFrame)
+            mlflow_payload_array = mlflow_payload.to_numpy()
```
Comment (sakoush) [inline runtimes/mlflow/mlserver_mlflow/encoding.py:38]: This is basically the crux of the PR, convert pandas -> numpy

File: runtimes/mlflow/mlserver_mlflow/encoding.py [hunk 3]
```diff
@@ -1,23 +1,48 @@
...
+        elif isinstance(mlflow_payload, pd.Series):
+            assert isinstance(mlflow_payload, pd.Series)
+            mlflow_payload_array = mlflow_payload.to_numpy()
+        else:
```
Comment (sakoush) [inline runtimes/mlflow/mlserver_mlflow/encoding.py:42]: This is basically the crux of this PR, convert pandas -> numpy

File: runtimes/mlflow/requirements-dev.txt [hunk 4]
```diff
@@ -1 +1,2 @@
 scikit-learn
+torch==1.9.0
```
Comment (sakoush) [inline runtimes/mlflow/requirements-dev.txt:2]: this seems redundant with the main requirements/dev.txt

File: requirements/dev.txt [hunk 5]
```diff
@@ -7,6 +7,10 @@ pytest==6.2.2
 pytest-asyncio==0.15.1
 pytest-mock==3.6.1
 tox==3.24.1
+# for mlflow tests
+torch==1.9.0
```
Comment (adriangonz) [inline requirements/dev.txt:11]: Since this is only required for the `mlserver-mlflow` package, should we remove them from this one?

File: runtimes/mlflow/tests/test_encoding.py (hunk 1 above)
Comment (adriangonz) [reply runtimes/mlflow/tests/test_encoding.py:19]: Mmmh good question... should we add some for multi-column dataframes maybe?

File: requirements/dev.txt (hunk 5 above)
Comment (sakoush) [reply requirements/dev.txt:11]: github tests only install from this list and it doesnt install the dev requirements for `mlserver-mlflow`

File: runtimes/mlflow/tests/test_encoding.py (hunk 1 above)
Comment (sakoush) [reply runtimes/mlflow/tests/test_encoding.py:19]: added

File: requirements/dev.txt (hunk 5 above)
Comment (adriangonz) [reply requirements/dev.txt:11]: Mmmh weird. Doesn't it run `make install-dev`? That one should install any `requirements-dev.txt` present on the runtime packages.

Comment (sakoush) [reply requirements/dev.txt]: it does yes and it should work, so removing these packages from here.

**Review by adriangonz — APPROVED** [review]
